File: utils/src/main/java/com/example/utils/AppLifecycleMonitor.java
package com.example.utils;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * 监听APP生命周期变化
 */
public class AppLifecycleMonitor implements Application.ActivityLifecycleCallbacks {

    //前后台状态回调监听器
    private ForegroundListener foregroundListener;
    //计数器
    private int count = 0;

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {

    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {
        synchronized (this) {
            if (count == 0) { //后台切换到前台
                if (foregroundListener != null) {
                    foregroundListener.getForegroundStatus(true);
                }
            }
            count++;
        }
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {

    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {

    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {
        synchronized (this) {
            count--;
            if (count == 0) { //前台切换到后台
                if (foregroundListener != null) {
                    foregroundListener.getForegroundStatus(false);
                }
            }
        }
    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {

    }

    public void setForegroundListener(ForegroundListener foregroundListener) {
        this.foregroundListener = foregroundListener;
    }

    public interface ForegroundListener {
        void getForegroundStatus(boolean foreground);
    }

}

File: library/utils/src/main/java/com/example/utils/AppLifecycleMonitor.kt
package com.example.utils

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle

/**
 * 监听APP生命周期变化
 */
class AppLifecycleMonitor : ActivityLifecycleCallbacks {

    //前后台状态回调监听器
    private var foregroundListener: ForegroundListener? = null

    //计数器
    private var count = 0
    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}

    override fun onActivityStarted(activity: Activity) {
        synchronized(this) {
            if (count == 0) { //后台切换到前台
                foregroundListener?.getForegroundStatus(true)
            }
            count++
        }
    }

    override fun onActivityResumed(activity: Activity) {}

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {
        synchronized(this) {
            count--
            if (count == 0) { //前台切换到后台
                foregroundListener?.getForegroundStatus(false)
            }
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}

    override fun onActivityDestroyed(activity: Activity) {}

    fun setForegroundListener(foregroundListener: ForegroundListener?) {
        this.foregroundListener = foregroundListener
    }

    interface ForegroundListener {
        fun getForegroundStatus(foreground: Boolean)
    }
}